Deprecate some variants of Get and MultiGet

anand1976 opened the pull request:
A lot of variants of Get and MultiGet have been added to `include/rocksdb/db.h` over the years. Try to consolidate them by marking variants that don't return timestamps as deprecated. The underlying DB implementation will check and return Status::NotSupported() if it doesn't support returning timestamps and the caller asks for it.

ltamasi commented on include/rocksdb/db.h:
```diff
@@ -549,53 +549,62 @@ class DB {
+
+  // DEPRECATED: This function might be removed in a future release
   virtual Status Get(const ReadOptions& options,
                      ColumnFamilyHandle* column_family, const Slice& key,
-                     PinnableSlice* value) = 0;
+                     PinnableSlice* value) final {
+    return Get(options, column_family, key, value, nullptr);
+  }
+
+  // DEPRECATED: This function might be removed in a future release
   virtual Status Get(const ReadOptions& options, const Slice& key,
-                     std::string* value) {
+                     std::string* value) final {
     return Get(options, DefaultColumnFamily(), key, value);
   }
```
Thanks for working on cleaning up all these overrides! I think changing the fallback logic so we implement older/less general APIs in terms of newer/more general ones, and having the fallback logic itself in the base class/interface (as opposed to repeating it in derived classes) makes perfect sense. There are few things I'm wondering about though:
1) Not sure if we would actually want to deprecate/remove the timestamp-less read APIs. From a theoretical perspective, user-defined timestamps are an optional part of our data model, and from a practical standpoint, even if we wanted to remove these APIs, migrating all existing users might not be realistic.
2) I think we might have some lower hanging fruit where a newer API provides a clearly better experience across the board: e.g. we might consider deprecating/removing the old-style non-batched `MultiGet` altogether, and potentially the point lookup variants that don't take a `ColumnFamilyHandle` and the ones that take `string` output parameters (although the latter two might be also in widespread usage).
3) On the technical side, our fallback methods in general could be regular non-virtual functions (instead `virtual` + `final`). This would also enable keeping the interface while eliminating the possibility of derived classes overriding some but not all variants.

pdillinger replied:
I'm also reluctant to deprecate functions that well represent a common subset of the data model. There's nothing wrong with convenience wrappers, especially if we are sure enough about their meaning in terms of other functions that we eliminate the option to override them.

Re: 3: See https://stackoverflow.com/a/11704714
> One reason some give for a base class to declare a non-overriding method to be final is simply so that anyone trying to define that method in a derived class gets an error instead of silently creating a method that 'hides' the base class's method.

Given that it was previously overridable, `virtual final` seems like the clear choice even though its weird at first glance.

anand1976 replied:
> Not sure if we would actually want to deprecate/remove the timestamp-less read APIs. From a theoretical perspective, user-defined timestamps are an optional part of our data model, and from a practical standpoint, even if we wanted to remove these APIs, migrating all existing users might not be realistic.

> I'm also reluctant to deprecate functions that well represent a common subset of the data model. There's nothing wrong with convenience wrappers, especially if we are sure enough about their meaning in terms of other functions that we eliminate the option to override them.

Makes sense. I went back and forth on these, but I agree there's some utility to having the timestamp-less variants in the user facing API. I'll not mark them as deprecated, but leave them as `final` so they cannot be overridden.

anand1976 replied:
Regarding declaring a function as `virtual final` vs non-virtual, is there a generally prescribed way of doing it? I guess its six of one half a dozen of the other. Personally, I like `virtual final` as it makes it pretty obvious it should not be overridden.

ltamasi commented on include/rocksdb/db.h:
```diff
@@ -705,33 +717,42 @@ class DB {
-                        const size_t num_keys, const Slice* keys,
-                        PinnableSlice* values, Status* statuses,
-                        const bool /*sorted_input*/ = false) {
+  virtual void MultiGet(const ReadOptions& options, const size_t num_keys,
+                        ColumnFamilyHandle** column_families, const Slice* keys,
+                        PinnableSlice* values, std::string* timestamps,
+                        Status* statuses, const bool /*sorted_input*/ = false) {
```
These overloads are kind of confusing (which is I guess why we're cleaning them up ;) ) and I might be getting lost here but isn't this the most general overload that others should be implemented in terms of (either directly or indirectly)?

ltamasi replied:
Oh, I see it falls back to the non-batched `MultiGet` logic! Again, I think we could consider deprecating those, or if not, consider flipping things around and implementing the older APIs in terms of the newer/better ones

anand1976 replied:
Not every derived class implements the batched `MultiGet`. In fact, only `DBImpl` implements it, hence we can't deprecate/redirect the non-batched one to the batched one.

pdillinger commented on include/rocksdb/db.h:
```diff
@@ -705,33 +717,42 @@ class DB {
   // keys - Pointer to C style array of key Slices with num_keys elements
   // values - Pointer to C style array of PinnableSlices with num_keys elements
+  // timestamps - Pointer to C style of std::string that, if non-null and
+  //              timestamps are enabled, will have the timestamps of the keys
```
Rule 0 of function documentation: what are the inputs and what are the outputs? It's not clear whether timestamps are inputs, outputs, or both. Probably not *just* input because not const, but not obvious.

anand1976 reviewed (commented):
Thanks for the review @ltamasi and @pdillinger

pdillinger commented:
> `virtual final`

You could add something like

```
// NOTE: virtual final => disallow override (was previously allowed)
```

anand1976 commented:
@ltamasi @pdillinger Another pass at this. Changes are -
1. Only Get/MultiGet variants that do not take a column family (i.e default column family) are marked deprecated. I think 6 minor variations of Get are too many (2 of which are now deprecated), but let me know if you think otherwise and we should keep those around for convenience.
2. Derived classes are required to implement one Get and one MultiGet function, and the rest are marked as `virtual final` and reimplemented in terms of the two required ones. The required functions take the timestamp(s) parameter, and concrete implementations return `Status::NotSupported` when necessary.
3. The non-batched MultiGet is now implemented using the batched version, since it didn't have any additional functionality and can easily be implemented using the batched version.
4. Updated the API comments to clarify usage, as well as requirements for the derived classes.

pdillinger commented:
All sounds good except

> 1. Only Get/MultiGet variants that do not take a column family (i.e default column family) are marked deprecated. I think 6 minor variations of Get are too many (2 of which are now deprecated), but let me know if you think otherwise and we should keep those around for convenience.

I think it's common enough to use a single CF that the versions that elide CF really are convenient. Sure, there's a tiny risk that someone forgets to specify which CF their data is supposed to read from / written to, but IMHO that's mostly an "I didn't test my code and expected it to work" kind of a risk.

I don't feel super strongly about it, but that's my opinion.

anand1976 commented:
> 
> I think it's common enough to use a single CF that the versions that elide CF really are convenient. Sure, there's a tiny risk that someone forgets to specify which CF their data is supposed to read from / written to, but IMHO that's mostly an "I didn't test my code and expected it to work" kind of a risk.
> 
> I don't feel super strongly about it, but that's my opinion.

Good point. Default CF usage is quite common. We could treat a nullptr passed to the CF version of Get/MultiGet as the default CF, though it would put the onus on implementations to check and replace. There's also the practical matter of forcing clients to change their code. So I guess I'll leave them alone for now and we can revisit later if appropriate.

pdillinger commented on db/db_impl/db_impl.cc:
```diff
@@ -3122,38 +2865,29 @@ void DBImpl::PrepareMultiGetKeys(
+                  std::string* timestamps, Status* statuses,
+                  const bool sorted_input) {
+  // Use an autovector, if possible, to avoid memory allocation overhead
+  autovector<ColumnFamilyHandle*, MultiGetContext::MAX_BATCH_SIZE> cf_autovec;
```
Hmm. autovector has a vector which is essentially unused here (could have been an array). Also, constructing + destructing vector is non-trivial when not needed. There's at least 3 lines of code dedicated to not duplicating a two-line call to other MultiGet(), so I'd suggest that duplicating the call is simpler and more efficient. (Save unnecessary construction / destruction / stack allocation.)

pdillinger commented on db/db_impl/db_impl.cc:
```diff
@@ -3122,38 +2865,29 @@ void DBImpl::PrepareMultiGetKeys(
-  if (read_options.io_activity == Env::IOActivity::kUnknown) {
-    read_options.io_activity = Env::IOActivity::kMultiGet;
+  for (size_t i = 0; i < num_keys; ++i) {
+    column_families[i] = column_family;
```
You should be able to populate std::vector or std::array with copies of a single value without the need for an explicit loop.

pdillinger reviewed (approved):
Thanks for the clean up!

Please update PR title and description.

(Re-)consider deprecating MultiGets returning `std::vector<Status>`?

anand1976 commented:
> 
> (Re-)consider deprecating MultiGets returning `std::vector<Status>`?

These return the values as `std::string` vs `PinnableSlice`. Its a little more convenient for users who are not performance sensitive. Given we're retaining convenience wrappers for default column family, and string version of Get, it doesn't seem like a big deal to keep these around IMO.

jaykorean commented on utilities/ttl/db_ttl_impl.cc:
```diff
@@ -505,22 +509,35 @@ Status DBWithTTLImpl::Get(const ReadOptions& options,
+    statuses[i] = StripTS(&values[i]);
   }
-  return statuses;
+  return;
```
no longer needed

anand1976 commented on db/db_impl/db_impl.cc:
```diff
@@ -2497,250 +2491,6 @@ Status DBImpl::GetImpl(const ReadOptions& read_options, const Slice& key,
-                  /*timestamps=*/nullptr);
-}
-
-std::vector<Status> DBImpl::MultiGet(
```
Caught a minor bug - missing stat update in the batched MultiGet code path. Merging the two types of MultiGet is already paying dividends ;)

anand1976 commented on db/db_basic_test.cc:
```diff
@@ -4504,68 +4504,35 @@ TEST_P(DBBasicTestMultiGetDeadline, MultiGetDeadlineExceeded) {
   // Delay the first IO
   fs->SetDelayTrigger(ro.deadline, ro.io_timeout, 0);
 
-  std::vector<Status> statuses = dbfull()->MultiGet(ro, cfs, keys, &values);
```
There are other tests for the non-batched MultiGet API itself. This test is for the deadline feature, which should be fully covered below since the the MultiGets now go through the same code path.
